File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.ts
class TrieNode<T> {
  public children: Map<T, TrieNode<T>>;
  public isEnd: boolean;

  constructor() {
    this.children = new Map<T, TrieNode<T>>();
    this.isEnd = false;
  }
}

class Trie {
  private root: TrieNode<string>;

  constructor() {
    this.root = new TrieNode<string>();
  }

  insert(word: string): void {
    let curr = this.root;
    for (let c of word) {
      if (!curr.children.has(c)) {
        curr.children.set(c, new TrieNode<string>());
      }
      curr = curr.children.get(c);
    }
    curr.isEnd = true;
  }

  search(word: string): boolean {
    let curr = this.root;
    for (let c of word) {
      if (!curr.children.has(c)) 
        return false;

      curr = curr.children.get(c);
    }
    return curr.isEnd;
  }

  startsWith(prefix: string): boolean {
    let curr = this.root;
    for (let c of prefix) {
      if (!curr.children.has(c))
        return false;

      curr = curr.children.get(c);
    }
    return true;
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.java
public class Trie {
  
  private class TrieNode {

    TrieNode[] next = new TrieNode[26];
    boolean isEnd = false;

    public TrieNode() { }

    public TrieNode(boolean isEnd) {
      this.isEnd = isEnd;
    }
    
  }
  
	TrieNode root;

	public Trie() {
		this.root = new TrieNode();
	}

	public void insert(String word) {
		TrieNode curr = this.root;

		for (char c : word.toCharArray()) {
			if (curr.next[c - 'a'] == null) {
				curr.next[c - 'a'] = new TrieNode();
			}
			curr = curr.next[c - 'a'];
		}

		curr.isEnd = true;
	}

	public boolean search(String word) {
		TrieNode curr = this.root;

		for (char c : word.toCharArray()) {
			if (curr.next[c - 'a'] == null) {
				return false;
			}
			curr = curr.next[c - 'a'];
		}

		return curr.isEnd;
	}

	public boolean startsWith(String prefix) {
		TrieNode curr = this.root;

		for (char c : prefix.toCharArray()) {
			if (curr.next[c - 'a'] == null) {
				return false;
			}
			curr = curr.next[c - 'a'];
		}

		return true;
	}
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */